lib/tests/triples.nix

```diff
@@ -0,0 +1,157 @@
+# Run:
+# [nixpkgs]$ nix-build -o expected-output lib/tests/triples.nix -A expected-output-from-gnu-config
+# [nixpkgs]$ nix-instantiate --eval --strict -E '(import lib/tests/triples.nix).tests ./expected-output'
+# Expected output: [], or the failed cases
+# Two steps are required to avoid IFD.
+#
+# This file tests our triple-parser, but it needs access to `pkgs`
+# because `pkgs.gnu-config` is our source of truth for the
+# correctness of parsing.  This file does not test `pkgs`; it tests
+# `lib`.
+#
+# The tests below attempt to verify that parsing followed by
+# unparsing behaves the same way as `gnu-config`, except in the
+# rare and unusual cases where nixpkgs has carefully and
+# deliberately decided to diverge from it.
+#
+# We use `gnu-config` as our source of ground truth not because it
+# is elegant (it isn't) or sensible (it isn't), but because it:
+#
+# 1. has an executable (i.e. unambiguous) specification
+# 2. covers a superset of the systems on which nixpkgs could be useful
+# 3. is *not* under nixpkgs' control (bickering/drama avoidance)
+# 4. emanates from an organization (GNU) which understands that
+#    the taxonomy is not merely an implementation detail of one of
+#    their software products, and has demonstrated this
+#    understanding over an extended period of time (multiple
+#    decades).
+#
+
+let
+  lib  = import ../default.nix;
+  pkgs = import ../.. {};
+
+  inherit (lib.systems.parse) canonicalize;
+
+  filtered-triples = lib.pipe null [
+
+    # start with the lists of known possibilities for each field
+    (_: with lib.systems.parse; {
+      cpu    = cpuTypes;
+      vendor = vendors;
+      kernel = kernels;
+      abi    = abis;
+    })
```

Comment:
@Ericson2314 here is where the tests exhaustively enumerate the space of triples to be tested.  `examples.nix` is not involved in any way, shape, or form.

Reply:
OK yes I agree this is better!

Reply:
> OK yes I agree this is better than going off `examples.nix`!

Which is now ready for deprecation and removal:

- https://github.com/NixOS/nixpkgs/pull/250073